### PR TITLE
WT-7514 Let tiered subsystem open files on behalf of block manager.

### PR DESCRIPTION
**Pull request by ddanderson**
I've based this branch off of wt-7499 changes, which have not yet been merged.
It shows how I expect the block manager to open files based on their object id.
This does not yet addressed flush_tier changes for the block manager.

### CODE REVIEW

**Comment by ddanderson** [comment]
@sueloverso, @keitharnoldsmith I've made this a draft since it's based off the wt-7499 (unmerged) pull request.  Here's the direction I'm going with the block manager to let the tier code have complete knowledge of how to name files and open them.

File: src/tiered/tiered_opener.c
Comment (sueloverso) [inline src/tiered/tiered_opener.c:26]: I don't know if you looked at it but the function that exists that you might want to use (maybe with some small changes) is `tiered/tiered_handle.c:__wt_tiered_name`.

File: src/tiered/tiered_handle.c
Comment (sueloverso) [inline src/tiered/tiered_handle.c:395]: I think we may want a new no-prefix name option to `wt_tiered_name`, maybe `WT_TIERED_NAME_ONLY` that doesn't prepend the prefix. With the changes I merged today for the new suffix, `prev_filename` and `new_object_name` should be identical always. The same object names and the prefix indicates where it lives. Does that mean `ss_flush*` should only have one name arg instead of two?

File: src/tiered/tiered_handle.c
Comment (sueloverso) [inline src/tiered/tiered_handle.c:383]: I pushed a draft branch of the thread and work unit stuff. It is still very much in progress but I'll move this code into it when I write the function to process a flush work unit.

**Comment by sueloverso** [comment]
I took a very quick look at this and left a couple comments.

**Comment by ddanderson** [comment]
@sueloverso and @keitharnoldsmith, I made some changes to treat `WT_DHANDLE_TYPE_TIER` as a btree in several more cases, that has resolved the "clean close connection" issue.  The tiered tests run cleanly.  If I add in a connection reopen to the test, it does not work yet, that's next.

**Review by keitharnoldsmith — COMMENTED** [review]
Nice work @ddanderson!

Glad you found that `WT_DHANDLE_TYPE_BTREE` check in `__wt_checkpoint_get_handles()`. That was probably important for getting data properly written.

File: src/tiered/tiered_opener.c
Comment (ddanderson) [reply src/tiered/tiered_opener.c:26]: Done.

**Comment by ddanderson** [comment]
@sueloverso and @keitharnoldsmith, this branch is ready to review again.  The principal goals of this PR are:

- advance the tests so that we can do `flush_tier` and see objects move correctly (e.g. into the `mybucket` and `cache-mybucket` directory, for test_tiered02.py).
- Identify some limits of testing (new tickets WT-7589, WT-7591) so it's clear what to work on next to extend the work.
- Identify some other code deficiencies with other new tickets (WT-7588, WT-7590)
- remove unused/dead code (some vestiges from LABS-19, like tracking the current object id in the block manager, etc.)

Some of the more significant parts of this change:

- The "file opener" concept
- New config option for files: `"tiered_object=true"`, that will prevent the file's Btree from writing out blocks that step on the tiered's Btree.  Getting the config strings copied too and fro to make this work is pretty ugly, I'll probably need to add some comments or (better) find a way to clean it up.
- Btree's and Tiered dhandles are looking even more alike, there are probably some opportunities for better code sharing.

File: src/tiered/tiered_handle.c
Comment (keitharnoldsmith) [inline src/tiered/tiered_handle.c:123]: Instead of reaching into the btree to figure out if the block manager has been created, could we let `__wt_btree_switch_object()` handle this?  

Alternately we could add a flag on `WT_TIERED` to indicate that we are initializing. Or perhaps the inverse, a flag that says our BTREE is setup.

**Review by keitharnoldsmith — APPROVED** [review]
LGTM.

One comment about possibly simplifying the initial case in `__tiered_create_local()`. 

There are a few things that feel a bit awkward, but they're already tagged and ticketed for future work.  So LGTM!

File: src/btree/bt_handle.c
Comment (sueloverso) [inline src/btree/bt_handle.c:401]: Further down is the `readonly` configuration that should be removed/modified/whatever once you decide which of these we need.

File: src/include/block.h
Comment (sueloverso) [inline src/include/block.h:321]: I find `TOP` a confusing term. Maybe `WT_TIERED_OBJECT_ID_MAX`? Or can this just be defined as `UINT64_MAX` - does it really need its own define or could we just use one of the existing ones?

File: src/tiered/tiered_opener.c
Comment (sueloverso) [inline src/tiered/tiered_opener.c:40]: The `wt_tiered_name` function could take a (new) `WT_TIERED_NAME_ONLY` flag perhaps (or `WT_TIERED_NAME_NO_PREFIX`). Callers in `conn_tiered.c:wt_tiered_do_flush` are also using it in this manner to the the URI and then stripping off the prefix. That could be another ticket, just pointing it out.

**Review by sueloverso — APPROVED** [review]
LGTM - I have a few comments.

**Comment by sueloverso** [comment]
I took a look and the failing task is `tiered04`. That should be fixed before merge. If it was random known problem it would be okay. But `tiered04` could be real.

File: dist/api_data.py
Comment (ddanderson) [reply dist/api_data.py:278]: Yes - the extra functionality currently is no checkpoint.  It's also likely that we shouldn't allow opening of this object on its own even as readonly (after all, it may have references to other objects, and the Block manager won't know what to do without viewpoint offered when opening "tiered:").  I'll remove readonly.

Comment (ddanderson) [reply dist/api_data.py:278]: As it turns out, we'll need readonly independently, as `"object:"` uris "inherit" from `"file:"` and objects need readonly.  So I'm going to leave this as is.

File: src/include/block.h
Comment (ddanderson) [reply src/include/block.h:321]: I named this `WT_TIERED_CURRENT_ID` and added a comment, it's a way in the opener API to reference the current id.

File: dist/api_data.py
Comment (sueloverso) [reply dist/api_data.py:278]: Thanks for researching it.

**Comment by ddanderson** [comment]
tiered04 doesn't fail for me, but it seems to be failing consistently with the PR builds.  Perhaps it's a difference in diagnostic vs no-diagnostic builds

**Comment by ddanderson** [comment]
@sueloverso and @keitharnoldsmith, please take another look.  I believe I've addressed the PR comments.  The errors in tiered04 would appear to be caused by two threads both getting the dhandle exclusively.  I didn't change that, but now the worker thread that does the flush of the file retries on getting the dhandle.

File: src/conn/conn_tiered.c
Comment (sueloverso) [inline src/conn/conn_tiered.c:137]: Minor: alphabetize

File: src/conn/conn_tiered.c
Comment (sueloverso) [inline src/conn/conn_tiered.c:139]: Alphabetize on this line too

File: src/conn/conn_tiered.c
Comment (sueloverso) [inline src/conn/conn_tiered.c:185]: Minor: WT uses `msecs` or `msec` generally. And `retry`  or `retries` instead of tries.

File: src/conn/conn_tiered.c
Comment (sueloverso) [inline src/conn/conn_tiered.c:213]: I suggest adding this statistic. It is two additional lines, the definition in the `stat_data.py` file and one increment here. That doesn't seem worth a new ticket.

File: src/conn/conn_tiered.c
Comment (sueloverso) [inline src/conn/conn_tiered.c:205]: There is a `WT_RETRY_MAX` define used in a few places that is 10. I suggest using that.

**Review by sueloverso — APPROVED** [review]
I have a couple small comments about the new EBUSY code. But it's good to go after that.

File: src/conn/conn_tiered.c
Comment (keitharnoldsmith) [reply src/conn/conn_tiered.c:205]: A potential problem with letting the number of (re)tries go to 10 is that the last sleep would be 100 seconds, which seems rather long for this type of race.  Obviously you could start with a shorter sleep, or keep the current retry count.  Either way it seems like guesswork until we start using the system for real and see how often this race happens and how long it lasts.

**Comment by keitharnoldsmith** [comment]
LGTM

File: src/conn/conn_tiered.c
Comment (sueloverso) [reply src/conn/conn_tiered.c:205]: I would probably adjust the multiplier first. Try the 10 times at a fixed, short interval.

File: src/btree/bt_handle.c
Comment (ddanderson) [reply src/btree/bt_handle.c:120]: Done.

File: src/block/block_tiered.c
Comment (ddanderson) [reply src/block/block_tiered.c:48]: Done.

File: src/tiered/tiered_handle.c
Comment (ddanderson) [reply src/tiered/tiered_handle.c]: I rewrote as you suggested.

File: src/conn/conn_tiered.c
Comment (ddanderson) [reply src/conn/conn_tiered.c:185]: Fixed.

File: src/conn/conn_tiered.c
Comment (ddanderson) [reply src/conn/conn_tiered.c:213]: Done.

File: src/conn/conn_tiered.c
Comment (ddanderson) [reply src/conn/conn_tiered.c:205]: Since msecs is doubling each time, I'm a little wary of bumping to 10, that makes the last wait 50 seconds.  I changed the code to exit the loop when the msecs > a few seconds, matching the comment.

File: src/tiered/tiered_opener.c
Comment (ddanderson) [reply src/tiered/tiered_opener.c:40]: I created WT-7635 to track this.

**Pull request merged**
